File: app/schemas/execution.py
from pydantic import BaseModel
from typing import Optional, Dict
from datetime import datetime
import uuid


class ExecutionBase(BaseModel):
    resource_id: uuid.UUID
    action: str
    desired_state: str
    actual_state_before: str
    status: str = "completed"  # pending, completed, failed
    success: bool
    error_message: Optional[str] = None


class Execution(ExecutionBase):
    id: uuid.UUID
    organization_id: uuid.UUID
    executed_at: datetime

    class Config:
        from_attributes = True


class ReconciliationResult(BaseModel):
    """Result of a manual reconciliation run."""
    processed: int
    actions_taken: int
    errors: int


class ExecutionStatistics(BaseModel):
    """Aggregated execution statistics for a time period."""
    total_executions: int
    successful: int
    failed: int
    success_rate: float
    actions: Dict[str, int]  # {"START": count, "STOP": count}
    period_hours: int


class ExecutionSummary(BaseModel):
    """Dashboard-style execution summary."""
    last_hour: ExecutionStatistics
    last_24_hours: ExecutionStatistics
    last_7_days: ExecutionStatistics
    recent_failures_count: int
    last_execution_at: Optional[datetime] = None

File: frontend/src/lib/types/execution.ts
export interface Execution {
  id: string; // UUID
  organization_id: string; // UUID
  resource_id: string; // UUID
  action: string; // "START" | "STOP"
  desired_state: string;
  actual_state_before: string;
  status: string; // "pending" | "completed" | "failed"
  success: boolean;
  error_message: string | null;
  executed_at: string; // ISO datetime
}

export interface ExecutionStatistics {
  total_executions: number;
  successful: number;
  failed: number;
  success_rate: number; // 0.0 - 1.0
  actions: Record<string, number>; // {"START": count, "STOP": count}
  period_hours: number;
}

export interface ExecutionSummary {
  last_hour: ExecutionStatistics;
  last_24_hours: ExecutionStatistics;
  last_7_days: ExecutionStatistics;
  recent_failures_count: number;
  last_execution_at: string | null; // ISO datetime
}

export interface ExecutionFilter {
  resource_id?: string;
  action?: string;
  success?: boolean;
  status?: string;
  hours?: number;
  limit?: number;
  offset?: number;
}
